ddtrace/internal/_encoding.pyx

```diff
@@ -278,13 +278,28 @@ cdef class MsgpackStringTable(StringTable):
             self.pk.length = self._sp_len
             self._next_id = self._sp_id
 
+        # After rolling back the string table next_id we must remove all stale string -> _id pairs
+        # This will resolve two classes of encoding errors:
+        #  - multiple strings referencing the same string table index. In this scenario
+        #    two different strings in the encoded trace could be serialized with the same _id. In this scenario 
+        #    two different strings could reference one string in the encoded trace (string swapping).
+        # - when the string table references an index of the string table that is not serialized. The encoded 
+        #    trace can not be decoded without accessing an invalid index. In this scenario the agent will 
+        #    return a 400 status code.
+        for string in list(self._table.keys()):
+            if self._table[string] >= self._next_id:
+                del self._table[string]
```

Comment:
we could probably also do a dict comprehension here too:

```python
self._table = {s: idx for s, idx in self._table.items() if idx < self._next_id}
```

just an observation, not sure it is any cleaning. could save on the extra data structures (the list) and operations (all the `del`s), but probably splitting hairs here.